rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+    BadLibraryVersion(String, String),
+    #[error("The Stan library was compiled without threading support. Config was {0}")]
+    StanThreads(String),
+    #[error("Failed to decode string to UTF8")]
+    InvalidString(#[from] Utf8Error),
+    #[error("Failed to construct model: {0}")]
+    ConstructFailed(String),
+    #[error("Failed during evaluation: {0}")]
+    EvaluationFailed(String),
+}
+
+type Result<T> = std::result::Result<T, BridgeStanError>;
+
+/// Open a compiled stan library.
+///
+/// The library should have been compiled with bridgestan,
+/// with the same version as the rust library.
+pub fn open_library<P: AsRef<OsStr>>(path: P) -> Result<StanLibrary> {
+    let library = unsafe { libloading::Library::new(path) }?;
+    let major: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_major_version") }?;
+    let major = unsafe { **major };
+    let minor: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_minor_version") }?;
+    let minor = unsafe { **minor };
+    let patch: libloading::Symbol<*const c_int> = unsafe { library.get(b"bs_patch_version") }?;
+    let patch = unsafe { **patch };
+    let self_major: c_int = env!("CARGO_PKG_VERSION_MAJOR").parse().unwrap();
+    let self_minor: c_int = env!("CARGO_PKG_VERSION_MINOR").parse().unwrap();
+    let self_patch: c_int = env!("CARGO_PKG_VERSION_PATCH").parse().unwrap();
+
+    if (self_major != major) | (self_minor != minor) {
+        return Err(BridgeStanError::BadLibraryVersion(
+            format!("{}.{}.{}", major, minor, patch),
+            format!("{}.{}.{}", self_major, self_minor, self_patch),
+        ));
+    }
+
+    let lib = unsafe { ffi::Bridgestan::from_library(library) }?;
+    let lib = ManuallyDrop::new(lib);
+    Ok(StanLibrary(lib))
+}
+
+/// A Stan model instance with data
+pub struct Model<T: Borrow<StanLibrary>> {
+    model: NonNull<ffi::bs_model>,
+    lib: T,
+}
+
+// Stan model is thread safe
+unsafe impl<T: Sync + Borrow<StanLibrary>> Sync for Model<T> {}
+unsafe impl<T: Send + Borrow<StanLibrary>> Send for Model<T> {}
+
+/// A random number generator for Stan
+pub struct Rng<T: Borrow<StanLibrary>> {
+    rng: NonNull<ffi::bs_rng>,
+    lib: T,
+}
+
+unsafe impl<T: Sync + Borrow<StanLibrary>> Sync for Rng<T> {}
+unsafe impl<T: Send + Borrow<StanLibrary>> Send for Rng<T> {}
+
+impl<T: Borrow<StanLibrary>> Drop for Rng<T> {
+    fn drop(&mut self) {
+        unsafe {
+            // We don't handle error messages during deconstruct
+            self.lib.borrow().0.bs_rng_destruct(self.rng.as_ptr());
+        }
+    }
+}
+
+impl<T: Borrow<StanLibrary>> Rng<T> {
+    pub fn new(lib: T, seed: u32) -> Result<Self> {
+        let mut err = ErrorMsg::new(lib.borrow());
+        let rng = unsafe {
+            lib.borrow()
+                .0
+                .bs_rng_construct(seed as c_uint, err.as_ptr())
+        };
+        if let Some(rng) = NonNull::new(rng) {
+            drop(err);
```

Comment:
We don't manually drop the error anywhere else, and I don't think it is necessary? But either way we should be consistent

Reply:
We need this drop explicitly in this particular case because we are only allowed to touch `lib` after `err` is dropped. And without the explicit drop it would only do that later.